### PR TITLE
[MAINTENANCE] Refactor `DataAssistant` plotting to leverage utility dataclasses

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- We often pair together a name, title, and type in our plotting logic - as such, it makes sense to group these fields together into dataclasses. Since these pairings share behavior, we can use an abstract class to define what behavior each of these dataclasses needs to implement.
  - Share behavior includes tooltip creation, axis creation, etc

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: great_expectations/rule_based_profiler/types/data_assistant_result/plot_util.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/plot_util.py]: There's perhaps a better name for this 😄 suggestions are welcomed.

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py]: Pattern I'm proposing works as follows - public methods take in the raw arguments, formats them into these utility classes, and passes them downstream to a private method to perform the actual logic. This keeps our API consistent and removes the need for the user to know about these dataclasses.

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py:269]: If we like the pattern used herein, I'm happy to apply it to `min_value` and `max_value` in this PR.

File: great_expectations/rule_based_profiler/types/data_assistant_result/plot_util.py
Comment (cdkini) [reply great_expectations/rule_based_profiler/types/data_assistant_result/plot_util.py]: Let's also block this PR until we get docstrings - I'm holding off on documentation until we deem this approach appropriate.

Edit: Just kidding the docstring checker won't let me merge even if I wanted to...

File: great_expectations/rule_based_profiler/types/data_assistant_result/plot_util.py
Comment (cdkini) [inline great_expectations/rule_based_profiler/types/data_assistant_result/plot_util.py]: Pros/cons of dataclasses here? These were originally larger value bags but it may make less sense now?

**Review by alexsherstinsky — APPROVED** [review]
LGTM (once Azure checks pass and others' feedback is incorporated).

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (NathanFarmer) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py]: Can we change this to `_get_expect_domain_values_to_be_between_chart` to make it clear that it works for more than one domain?

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (NathanFarmer) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py]: Maybe `plot_components` is the module name and classes are `BatchPlotComponent` (also removed the `Id`), `DomainPlotComponent`, `MetricPlotComponent`, and `ExpectationKwargPlotComponent`.

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (NathanFarmer) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py]: I believe what you are calling `shorthand` here is actually the required parameter `field`. The `shorthand` usually refers to a string with the field, and/or type, and/or aggregate (which you correctly described in the docstring for `plot_on_axis`). So if an actual shorthand is passed here I'm not sure what the override behavior is in the `alt.Y` object because you already set the type and title in the method using `self`.

Are we intending for the user to be able to override the Altair data type here? I think I understand why you use `metric_plot_util` and then override the name with `min_value_plot_util.name`, but why wouldn't we just use `y=min_value_plot_component.plot_on_axis()` here?

File: great_expectations/rule_based_profiler/types/data_assistant_result/plot_util.py
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/types/data_assistant_result/plot_util.py]: I'm happy with the dataclasses since it seems to clean up and simplify `DataAssistantResult`. They would probably be fine in the `data_assistant_result.py` module at the bottom though.

**Review by NathanFarmer — CHANGES_REQUESTED** [review]
Great ideas! I know this is WIP, but I have some comments on the state of this remote version.

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (cdkini) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py]: I agree! Done ✅

File: great_expectations/rule_based_profiler/types/data_assistant_result/plot_util.py
Comment (cdkini) [reply great_expectations/rule_based_profiler/types/data_assistant_result/plot_util.py]: I'm erring on the side of a separate file just in case the types expand but let me know if you feel strongly about consolidation.

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (cdkini) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py]: yes! good call

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (cdkini) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py]: ooh okay I think I misread the documentation I appreciate the clarification.

Hmm let me look into this I remember having some issues but I'll look again with fresh eyes.

Comment (cdkini) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py]: the docs seem to say it's the shorthand? https://altair-viz.github.io/user_guide/generated/channels/altair.X.html

Comment (cdkini) [reply great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py]: also the reason why I passed in that value instead of running `min_value_plot_component.plot_on_axis()` is because we're mixing metric and min_value attrs.

This was the previous code:
```
y=alt.Y(min_value, type=metric_type, title=metric_title),
```
the `type` and `title` correspond to the `metric_plot_component` but `min_value` has to come from `min_value_plot_component`.

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (NathanFarmer) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py]: Can `subtitle` become an optional property of `DomainPlotComponent`? Right now, a subtitle is needed for all domain types except `MetricDomainTypes.TABLE`.

**Review by NathanFarmer — CHANGES_REQUESTED** [review]
One requested change and then I think we are good to go!

File: great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py
Comment (NathanFarmer) [inline great_expectations/rule_based_profiler/types/data_assistant_result/data_assistant_result.py]: Shouldn't this be `field=min_value_component.name`? I'm struggling to follow the logic here. I feel like we should have a sub-class called `ExpectationKwargPlotComponent` that is used here for `min_value` and `max_value`.

**Review by NathanFarmer — CHANGES_REQUESTED** [review]
We missed updating `shorthand` to `field` and I'm still thinking it would be much clearer to have an `ExpectationKwargPlotComponent` instead of re-using pieces of `MetricPlotComponent` for plotting the kwargs.

**Review by NathanFarmer — APPROVED** [review]
LGTM!

**Pull request merged**
